File: linked_lists/return_kth_to_last/sol3.py
import sys

sys.path.append('../')

from sll import sll


def kth_to_last(slinked_list, k):
   nd = slinked_list.head
   
   nd_a = nd
   i = 0
   while i < k:
     nd_a = nd_a.next
     i += 1
     if not nd_a:
       if i == k:
         return nd.data
       else:
         return None
   
   while nd_a:
     nd_a = nd_a.next
     nd = nd.next
     
   return nd.data
  
  
if __name__ == "__main__":
  slinked_list = sll.sll()
  value = [1, 2, 3, 1, 3, 4, 5, 1, 2, 3]
  key = 0
  for it in value:
    slinked_list.addLast(key, it)
    key += 1
    
  print kth_to_last(slinked_list, 3)
  
 
 
 
 
 
 

File: linked_lists/return_kth_to_last/sol3.cpp
#include "sll.h"

#include <iostream>
#include <vector>

int kth_to_last(sll& slinked_list, int k) {
  node* nd = slinked_list.head();
  node* nd_a = nd->next();
  
  for (auto i = 1; i < k; ++i) {
    nd_a = nd_a->next();
  }
  
  while (nd_a) {
    nd = nd->next();
    nd_a = nd_a->next();
  }
  
  return nd->data();
}



int main() {
  sll slinked_list;
  std::vector<int> value{1, 2, 3, 1, 3, 4, 5, 1, 2, 3};
  int key = 0;
  for (auto& it : value) {
    std::cout << key << std::endl;
    if (key == 0) {
      slinked_list.insertHead(it, key);
    } else {
      slinked_list.insertLast(it, key);
    }
    ++key;
  }
  
  node* nd = slinked_list.head();
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
  
  std::cout << "getting result: ";
  int k;
  while (std::cin >> k) { 
   int res = kth_to_last(slinked_list, k);
   std::cout << "result is " << res << std::endl;
  }
}